### ISSUE TITLE
Content from SDL based apps is not shown over VNC in headless configuration in Debian 12

### ISSUE THREAD

**Issue by Srandista** [libsdl-org/SDL#8792]
I'm using Raspberry Pi 3B+ without a monitor attached (via VNC). However, in latest version of Raspberry Pi OS based on Debian 12 Bookworm, I found out, that SDL based games can't be played via VNC if no monitor is connected. At first, I was getting "No available displays" error like mentioned here #8054, but that was solved, when I upgraded to SDL 2.28.5 from Debian Sid.

But even with the latest SDL library, I'm unable to play any games via VNC without the local monitor connected or with the local monitor connected, but turned off, because the content of app is not being refreshed over VNC. When the local monitor is turned on, then it works as it supposed to, I can see and play the games without issues, but that defeats the purpose of headless configuration.

The issue only presents in Debian 12 Bookworm, in Debian 11 Bullseye it works without problems, with no monitor attached (or monitor turned off).

The platform is RPi 3B+ on aarch64 with Debian 12 (the issue appeared after upgrading from Debian 11).

First discovery is described on JA2 Sracciatella GitHub, since at first I thought it's only problem with that game, but later I found out, that it exist in any SDL based game: [https://github.com/ja2-stracciatella/ja2-stracciatella/issues/1949](url)

**Comment by slouken** [comment]
Can you investigate and see what changed at the OS level to prevent this from working? SDL doesn't have any VNC display support, so I'm not sure how this is supposed to work.

**Comment by icculus** [comment]
I have a Pi3 up and running with this Debian Bookworm image: https://raspi.debian.net/tested/20231109_raspi_3_bookworm.img.xz

What are you using for VNC? Do you have an X server running? Do _other_ things show up over VNC and SDL just doesn't show anything? Does it report an error instead of start the SDL app?

(And were you really using Raspberry Pi OS, which is based on Bookworm, but installs a desktop environment by default?)

**Comment by icculus** [comment]
Ah, failed to read carefully, it's raspberry pi OS.

**Comment by Srandista** [comment]
Yes, I'm using latest Raspberry Pi OS based on Bookworm, specifically this image: https://downloads.raspberrypi.com/raspios_arm64/images/raspios_arm64-2023-12-06/2023-12-05-raspios-bookworm-arm64.img.xz

I'm using the default desktop environment, iirc it's based on LXDE. In it, there's integrated RealVNC, which needs to be enabled first, but I also tried it with x11vnc and got the same symptoms.

Everything is being shown over VNC (like desktop, terminal, other apps), but when I run any SDL app (I tried 3 games, Vanilla Conquer, DevilutionX and Jagged Alliance 2 Stracciatella), I can see the window of an app, but its content is not being refreshed, so I can only see the black filled window (in case of Vanilla Conquer and DevilutionX) or flashing SirTech logo (which is the first thing shown, when running Jagged Alliance 2 Stracciatella). Also, no error is being shown, when using SDL 2.28.5.

**Comment by icculus** [comment]
Okay, do this when starting the game:

```bash
SDL_RENDER_DRIVER=software SDL_FRAMEBUFFER_ACCELERATION=false ./WhateverGameYouAreRunning
```

What appears to be happening is VNC can't transmit OpenGL rendering (either on the Pi or maybe just in general), and even for software rendered games, SDL will try to use OpenGL to blit to the screen more quickly by default.

If you set _both_ of those environment variables, SDL won't use OpenGL for rendering the 2D game (`SDL_RENDER_DRIVER`), and it won't try to use OpenGL to get the software-rendered bits to the screen faster.

Without those set, SDL's testsprite2 program experiences the exact problem you are describing. If both are set, I get rendering over VNC from a Pi 3 with the Raspberry Pi OS image you specified.

If the game itself uses OpenGL directly, instead of just using SDL's 2D rendering API, this won't help you either, but there are likely ways to convince Mesa to use software rendering, too.

**Comment by Srandista** [comment]
Thanks, this workaround seems to work for the mentioned games! Can I ask, were you able to determine, why this stopped working (and which OS component is responsible for the breakage) during transition from Debian 11 to Debian 12?

**Comment by icculus** [comment]
> Can I ask, were you able to determine, why this stopped working (and which OS component is responsible for the breakage) during transition from Debian 11 to Debian 12?

I have no idea, sorry.
